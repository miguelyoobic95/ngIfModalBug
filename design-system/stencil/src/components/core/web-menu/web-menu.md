---
name: Web Menu
category: Core Components
---

```web-menu.html
    <yoo-web-menu></yoo-web-menu>
```
```web-menu.js
    var menu = document.querySelector('yoo-web-menu');
    var entry = {
      logo: "assets/logo/operations_simple.svg",
      items:[{label: 'action1', icon:'yo-fire', handler: () => {}},
                {label: 'action2', icon:'yo-yo', handler: () => {}},
                {label: 'action3', icon:'yo-more', handler: () => {}, separator: true},
                {label: 'page1', icon:'yo-trophy', href:"http://trophy", badge: '8'},
                {label: 'page2', icon:'yo-star', href:"http://star", badge: '28'},
                {label: 'page3', icon:'yo-fire', href:"http://fire", badge: '568'}],
      user: {imgData: "https://resources.stuff.co.nz/content/dam/images/1/i/o/a/a/c/image.related.StuffLandscapeSixteenByNine.620x349.1ioalf.png/1492902717643.jpg"}
    };
  menu.entry = entry;
```

## Attributes

|Name|Type|Default|Description|
|---|---|---|---|
|`entry`|IWebMenuEntry|   |entry of the menu|

## CSS

|Type|Name|Description|
|---|---|---|


## Events
|Attr|Description|
|---|---|---|
|`itemClicked`|fired when an action is clicked, send the action|
|`profilClicked`|fired when the avatar user is clicked, send the user|