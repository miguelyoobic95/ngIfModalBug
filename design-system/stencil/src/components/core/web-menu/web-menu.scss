

:host {
    .menu-container {
        background-color: #f2f2f2;
        height: 100%;
        width: 68px;
        padding: 0.8rem;
        & > * {
            margin-left: 0.2rem;
        }
        .menu-logo {
            height: 36px;
            width: 36px;
            margin-bottom: 2rem;
            .image {
                height: 100%;
            }
        }

        .menu-items {
            margin-bottom: 1rem;

            .item-container {
                position: relative;
                &.item-separator {
                    margin-bottom: 4rem;
                }
            }

            .menu-icon {
                position: relative;
                border-radius: 50%;
                border: 1px solid cssvar(success);
                background-color: cssvar(light);
                height: 36px;
                width: 36px;
                margin-bottom: 1rem;
                i {
                    position: absolute;
                    color: cssvar(success);
                    top: 50%;
                    left: 50%;
                    transform: translate(-50%,-50%);
                }
    
                &:hover {
                    background: cssvar(gradient-success);
                    i {
                        color: cssvar(light);
                    }
                }
    
                &.selected {
                    background: cssvar(gradient-success);
                    i {
                        color: cssvar(light);
                    }
                }
                /deep/ yoo-badge {
                    .outer-container {
                        position: absolute;
                        top: -0.1875rem;
                        right: -0.25rem;
                    }
                }
            }

        }

        

        .menu-user {
            flex: 0;
            height: 100%;
            border-top: 1px solid cssvar(stable);
            margin-right: -0.8rem;
            margin-left: -0.8rem;
            padding-left: 1rem;
            padding-top: 1rem;
        }
    }
}