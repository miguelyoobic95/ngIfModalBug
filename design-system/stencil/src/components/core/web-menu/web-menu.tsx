import { Component, Element, Prop, Event, EventEmitter, State, Method } from '@stencil/core';
import { IWebMenuEntry } from '@shared/interfaces';
import { getBackImageStyle, cloudinary, resizeWindow } from '../../../utils/helpers'; //intersectionObserve

@Component({
    tag: 'yoo-web-menu',
    styleUrl: 'web-menu.scss',
    scoped: true
})
export class WebMenuComponent {

    @Prop() entry: IWebMenuEntry;
    @Prop() activePage: string; // name of the active page

    @Event() itemClicked: EventEmitter<any>;
    @Event() profilClicked: EventEmitter<any>;

     @Element() host: HTMLStencilElement;

    @State() logoWidth: number = 34;
    @State() logoHeight: number = 34;
    @State() scrollHeight: string;

    tooltipOptions: any = {
        theme: 'dark',
        size: 'small',
        distance: 14
    };

    isActivePage(page: any): boolean {
        return page.href && this.activePage && page.href === this.activePage;
    }

    @Method()
    resize() {
        let slim = this.host.querySelector('yoo-slim-scroll');
        if (slim) {
            this.scrollHeight = slim.clientHeight - (this.host.querySelector('.menu-container').clientHeight - window.innerHeight) + 'px';
        }
    }

    componentDidLoad() {
        setTimeout(() => this.resize(), 300);
        resizeWindow(() => this.resize());
    }

    renderUser(): JSX.Element {
        return (this.entry ?
            <yoo-avatar class="small" user={this.entry.user} onClick={() => this.profilClicked.emit(this.entry.user)}></yoo-avatar>
            : null);
    }

    render(): JSX.Element {
        return (
            <div class="menu-container" attr-layout="column">
                {this.entry && this.entry.logo ?
                    <div class="menu-logo">
                        <div class="image" style={getBackImageStyle(cloudinary(this.entry.logo, this.logoWidth, this.logoHeight))} />
                    </div> : null}
                <yoo-slim-scroll height={this.scrollHeight}>
                    <div>
                        {this.entry && this.entry.items && this.entry.items.length ?
                            <div class="menu-items" attr-layout="column">
                                {this.entry.items.map(a =>
                                    <div class={'item-container ' + (a.separator ? 'item-separator' : '')}>
                                        <div>
                                            <yoo-tooltip placement="left" text={a.label} options={this.tooltipOptions}>
                                                {a.icon ?
                                                    <div class={'menu-icon' + (this.isActivePage(a) ? ' selected' : '')}
                                                        onClick={() => { this.itemClicked.emit(a); }}>
                                                        {a.iconSelected && this.isActivePage(a) ?
                                                            <i class={a.iconSelected}></i>
                                                            : <i class={a.icon}></i>}
                                                        {a.badge ? <yoo-badge text={a.badge} class="notification danger"></yoo-badge> : null}
                                                    </div> : null}
                                            </yoo-tooltip>
                                        </div>
                                    </div>
                                )}
                            </div> : null}
                    </div>
                </yoo-slim-scroll>
                {this.entry && this.entry.user ?
                    <div class="menu-user">
                        {this.renderUser()}
                    </div> : null}
            </div>
        );
    }
}
