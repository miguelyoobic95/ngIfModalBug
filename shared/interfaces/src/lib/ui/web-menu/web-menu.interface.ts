import { IUser } from '../../entities/user/user.interface';

export interface IWebMenuEntry {
    logo?: string;
    items?: Array<{
        label: string;
        icon: string;
        iconSelected?: string;
        href?: string;
        name?: string
        handler?: Function;
        separator?: boolean;
        badge?: string;
    }>;
    user?: IUser;
}