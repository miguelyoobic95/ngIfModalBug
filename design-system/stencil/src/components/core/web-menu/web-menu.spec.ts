import { TestWindow } from '@stencil/core/testing';
import { WebMenuComponent } from './web-menu';

describe('WebMenuComponent', () => {
    it('should build', () => {
        expect(new WebMenuComponent()).toBeTruthy();
    });

    describe('Rendering', () => {
        let element, window;
        beforeEach(async () => {
            window = new TestWindow();
            element = await window.load({
                components: [WebMenuComponent],
                html: '<yoo-web-menu></yoo-web-menu>'
            });
        });

        it('Should render', async () => {
            expect(element).toMatchSnapshot();
        });

        it('Should render with an entry', async () => {
            let entry = {
                logo: 'assets/logo/operations_simple.svg',
                items: [{ label: 'action1', icon: 'yo-fire', handler: () => { } },
                { label: 'action2', icon: 'yo-yo', handler: () => { } },
                { label: 'action3', icon: 'yo-more', handler: () => { } },
                { label: 'page1', icon: 'yo-trophy', href: 'http://trophy' },
                { label: 'page2', icon: 'yo-star', href: 'http://star' },
                { label: 'page3', icon: 'yo-fire', href: 'http://fire' }],
                user: { imgData: 'https://resources.stuff.co.nz/content/dam/images/1/i/o/a/a/c/image.related.StuffLandscapeSixteenByNine.620x349.1ioalf.png/1492902717643.jpg' }
            };
            element.entry = entry;
            await window.flush();
            expect(element).toMatchSnapshot();
        });
    });

    describe('Events', () => {

        let emitEvent, window;
        beforeEach(async () => {
            window = new TestWindow();
            emitEvent = jest.fn();
        });

        it('Should emit an event when clicked on the avatar', async () => {
            let element = await window.load({
                components: [WebMenuComponent],
                html: '<yoo-web-menu></yoo-web-menu>'
            });
            let entry = { user: { imageData: 'http://dfghl' } };
            element.entry = entry;
            await window.flush();
            let avatar = element.querySelector('yoo-avatar');
            window.document.addEventListener('profilClicked', emitEvent);
            await avatar.click();
            expect(emitEvent).toHaveBeenCalled();
            expect(emitEvent).toHaveEventData(element.entry.user);
        });

        it('Should emit an event when clicked on an item', async () => {
            let element = await window.load({
                components: [WebMenuComponent],
                html: '<yoo-web-menu></yoo-web-menu>'
            });
            let handler1 = jest.fn();
            let entry = {
                items: [
                    { label: 'action1', icon: 'yo-fire', handler: handler1 },
                    { label: 'action2', icon: 'yo-yo', handler: () => { } },
                    { label: 'action3', icon: 'yo-more', handler: () => { } }
                ]
            };
            element.entry = entry;
            await window.flush();
            let item = element.querySelector('.menu-items .menu-icon');
            window.document.addEventListener('itemClicked', emitEvent);
            await item.click();
            expect(emitEvent).toHaveBeenCalled();
            expect(emitEvent).toHaveEventData(element.entry.items[0]);
        });
    });
});