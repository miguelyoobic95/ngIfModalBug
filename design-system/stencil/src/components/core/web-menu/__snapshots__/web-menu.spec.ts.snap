// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WebMenuComponent Rendering Should render 1`] = `
<yoo-web-menu
  class="hydrated"
  data-yoo-web-menu-host=""
>
  
  <div
    attr-layout="column"
    class="menu-container"
    data-yoo-web-menu=""
  >
    
    <yoo-slim-scroll
      data-yoo-web-menu=""
    >
      <div
        data-yoo-web-menu=""
      >
        
      </div>
    </yoo-slim-scroll>
    
  </div>
</yoo-web-menu>
`;

exports[`WebMenuComponent Rendering Should render with an entry 1`] = `
<yoo-web-menu
  class="hydrated"
  data-yoo-web-menu-host=""
>
  
  <div
    attr-layout="column"
    class="menu-container"
    data-yoo-web-menu=""
  >
    <div
      class="menu-logo"
      data-yoo-web-menu=""
    >
      <div
        class="image"
        data-yoo-web-menu=""
        style="background-repeat: no-repeat; background-size: cover; background-position-x: 5;0%; background-image: url(assets/logo/operations_simple.svg);"
      />
    </div>
    <yoo-slim-scroll
      data-yoo-web-menu=""
    >
      <div
        data-yoo-web-menu=""
      >
        <div
          attr-layout="column"
          class="menu-items"
          data-yoo-web-menu=""
        >
          <div
            class="item-container"
            data-yoo-web-menu=""
          >
            <div
              data-yoo-web-menu=""
            >
              <yoo-tooltip
                data-yoo-web-menu=""
                placement="left"
                text="action1"
              >
                <div
                  class="menu-icon"
                  data-yoo-web-menu=""
                >
                  <i
                    class="yo-fire"
                    data-yoo-web-menu=""
                  />
                  
                </div>
              </yoo-tooltip>
            </div>
          </div>
          <div
            class="item-container"
            data-yoo-web-menu=""
          >
            <div
              data-yoo-web-menu=""
            >
              <yoo-tooltip
                data-yoo-web-menu=""
                placement="left"
                text="action2"
              >
                <div
                  class="menu-icon"
                  data-yoo-web-menu=""
                >
                  <i
                    class="yo-yo"
                    data-yoo-web-menu=""
                  />
                  
                </div>
              </yoo-tooltip>
            </div>
          </div>
          <div
            class="item-container"
            data-yoo-web-menu=""
          >
            <div
              data-yoo-web-menu=""
            >
              <yoo-tooltip
                data-yoo-web-menu=""
                placement="left"
                text="action3"
              >
                <div
                  class="menu-icon"
                  data-yoo-web-menu=""
                >
                  <i
                    class="yo-more"
                    data-yoo-web-menu=""
                  />
                  
                </div>
              </yoo-tooltip>
            </div>
          </div>
          <div
            class="item-container"
            data-yoo-web-menu=""
          >
            <div
              data-yoo-web-menu=""
            >
              <yoo-tooltip
                data-yoo-web-menu=""
                placement="left"
                text="page1"
              >
                <div
                  class="menu-icon"
                  data-yoo-web-menu=""
                >
                  <i
                    class="yo-trophy"
                    data-yoo-web-menu=""
                  />
                  
                </div>
              </yoo-tooltip>
            </div>
          </div>
          <div
            class="item-container"
            data-yoo-web-menu=""
          >
            <div
              data-yoo-web-menu=""
            >
              <yoo-tooltip
                data-yoo-web-menu=""
                placement="left"
                text="page2"
              >
                <div
                  class="menu-icon"
                  data-yoo-web-menu=""
                >
                  <i
                    class="yo-star"
                    data-yoo-web-menu=""
                  />
                  
                </div>
              </yoo-tooltip>
            </div>
          </div>
          <div
            class="item-container"
            data-yoo-web-menu=""
          >
            <div
              data-yoo-web-menu=""
            >
              <yoo-tooltip
                data-yoo-web-menu=""
                placement="left"
                text="page3"
              >
                <div
                  class="menu-icon"
                  data-yoo-web-menu=""
                >
                  <i
                    class="yo-fire"
                    data-yoo-web-menu=""
                  />
                  
                </div>
              </yoo-tooltip>
            </div>
          </div>
        </div>
      </div>
    </yoo-slim-scroll>
    <div
      class="menu-user"
      data-yoo-web-menu=""
    >
      <yoo-avatar
        class="small"
        data-yoo-web-menu=""
      />
    </div>
  </div>
</yoo-web-menu>
`;
